src/threadinterrupt.h

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_THREADINTERRUPT_H
+#define BITCOIN_THREADINTERRUPT_H
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+
+/*
+    A helper class for interruptible sleeps. Calling operator() will interrupt
+    any current sleep, and after that point operator bool() will return true
+    until reset.
+*/
+class CThreadInterrupt
+{
+public:
+    explicit operator bool() const;
+    void operator()();
+    void reset();
+    bool sleep_for(std::chrono::milliseconds rel_time);
+    bool sleep_for(std::chrono::seconds rel_time);
+    bool sleep_for(std::chrono::minutes rel_time);
```

Comment:
Can we get rid of the duplication here and just declare `bool sleep_for(std::chrono::nanoseconds rel_time);`?

The conversion from other duration types would be automatic then, and we should be safe from overflows as long as we stay below 292 years as the maximum duration, but please, double check in case you decide to proceed this way.

Reply:
Even better, now `sleep_for` has been templetized, and accepts any `std::chrono::duration` type, leveraging `wait_for()` function the conversion across different duration types.